src/metaopt/io/parsing.py

```diff
@@ -0,0 +1,257 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""
+:mod:`metaopt.io.parsing` -- Functions for configuration parsing and resolving
+==============================================================================
+
+.. module:: parsing
+   :platform: Unix
+   :synopsis: How does metaopt resolve configuration settings?
+
+ - Experiment name resolves like this:
+   cmd-arg > cmd-provided moptconfig > `DEF_CMD_EXP_NAME`
+
+ - Database options resolve with the following precedence (high to low):
+   cmd-provided moptconfig > env vars > default files > defaults
+
+.. seealso:: `ENV_VARS`, `ENV_VARS_DB`
+
+ - All other managerial, `Optimization` or `Dynamic` options resolve like this:
+   cmd-args > cmd-provided moptconfig > database (if experiment name can be
+   found) > default files
+
+Default files are given as a list at `DEF_CONFIG_FILES_PATHS` and a precedence
+is respected when building the settings dictionary:
+default metaopt example file < system-wide config < user-wide config
+
+.. note:: `Optimization` entries are required, `Dynamic` entry is optional.
+
+"""
+from __future__ import absolute_import
+import os
+import socket
+import argparse
+import textwrap
+import logging
+from copy import deepcopy
+from collections import defaultdict
+
+import six
+import yaml
+
+import metaopt
+from metaopt import (optim, dynamic)
+
+# Define type of arbitrary nested defaultdicts
+nesteddict = lambda: defaultdict(nesteddict)
+
+log = logging.getLogger(__name__)
+
+################################################################################
+#                 Default Settings and Environmental Variables                 #
+################################################################################
+
+# Default settings for command line arguments (option, description)
+DEF_CMD_MAX_TRIALS = (-1, 'inf/until preempted')
+DEF_CMD_POOL_SIZE = (10, str(10))
+DEF_CMD_EXP_NAME = (None, '{user}_{starttime}')
+
+DEF_CONFIG_FILES_PATHS = [
+    os.path.join(metaopt.dirs.site_data_dir, 'moptconfig.yaml.example'),
+    os.path.join(metaopt.dirs.site_config_dir, 'moptconfig.yaml'),
+    os.path.join(metaopt.dirs.user_config_dir, 'moptconfig.yaml')
+    ]
+
+# list containing tuples of
+# (environmental variable names, configuration keys, default values)
+ENV_VARS_DB = [
+    ('METAOPT_DB_NAME', 'name', 'MetaOpt'),
+    ('METAOPT_DB_TYPE', 'type', 'MongoDB'),
```

Comment:
In the case `PostgreSQL` or `Mysql` would be chosen, this would represent the class name inside `metaopt/io/database`?